OMX-27-firmware/sequencer.h

```diff
@@ -60,71 +62,106 @@ struct PatternSettings {  // ?? bytes
-  { 15, 0, 0, 0, 0, 0, 1, 3, 1, 0, false, false, false, false },
-  { 15, 1, 0, 0, 0, 0, 1, 3, 1, 0, false, false, false, false },
-  { 15, 2, 0, 0, 0, 0, 1, 3, 1, 0, false, false, false, false },
-  { 15, 3, 0, 0, 0, 0, 1, 3, 1, 0, false, false, false, false },
-  { 15, 4, 0, 0, 0, 0, 1, 3, 1, 0, false, false, false, false },
-  { 15, 5, 0, 0, 0, 0, 1, 3, 1, 0, false, false, false, false },
-  { 15, 6, 0, 0, 0, 0, 1, 3, 1, 0, false, false, false, false },
-  { 15, 7, 0, 0, 0, 0, 1, 3, 1, 0, false, false, false, false }
-};
+  bool sendCV : 1;
 
-struct TimePerPattern {
-  Micros lastProcessTimeP : 32;
-  Micros nextStepTimeP : 32;
-  Micros lastStepTimeP : 32;
-  int lastPosP : 16;
-};
+  StepNote steps[NUM_STEPS]; // note data
+}; // ? bytes
 
-TimePerPattern timePerPattern[NUM_PATTERNS] = {
-  { 0, 0, 0 },
-  { 0, 0, 0 },
-  { 0, 0, 0 },
-  { 0, 0, 0 },
-  { 0, 0, 0 },
-  { 0, 0, 0 },
-  { 0, 0, 0 },
-  { 0, 0, 0 }
+// holds state for sequencer
+class SequencerState {
+
+public:
+  int ticks;                      // A tick of the clock
+  bool clockSource;               // Internal clock (0), external clock (1)
+  bool playing;                   // Are we playing?
+  bool paused;                    // Are we paused?
+  bool stopped;                   // Are we stopped? (Must init to 1)
+  byte songPosition;              // A place to store the current MIDI song position
+  int playingPattern;             // The currently playing pattern, 0-7
+  bool seqResetFlag;              // for autoreset functionality
+  int clockDivMult;               // TODO: per pattern setting
+  word stepCV;
+  int seq_velocity;
+  int seq_acc_velocity;
+
+  // TODO: move into Pattern?
+  int seqPos[NUM_PATTERNS]; // What position in the sequence are we in?
+
+  int patternDefaultNoteMap[NUM_PATTERNS]; // default to GM Drum Map for now
+  Pattern patterns[NUM_PATTERNS];
+
+  // TODO: move into Pattern?
+  TimePerPattern timePerPattern[NUM_PATTERNS];
+
+  Pattern* getPattern(int pattern) {
+    return &this->patterns[pattern];
+  }
+
+  Pattern* getCurrentPattern() {
+    return getPattern(this->playingPattern);
+  }
+
+  // Helpers to deal with 1-16 values for pattern length and channel when they're stored as 0-15
+  uint8_t getPatternLength(int pattern) {
+    return this->patterns[pattern].len + 1;
+  }
+
+  void setPatternLength(int pattern, int len) {
+    this->patterns[pattern].len = len - 1;
+  }
+
+  uint8_t getPatternChannel(int pattern) {
+    return this->patterns[pattern].channel + 1;
+  }
 };
 
-// Helpers to deal with 1-16 values for pattern length and channel when they're stored as 0-15
-uint8_t PatternLength( int pattern ) {
-  return patternSettings[pattern].len + 1;
-}
-
-void SetPatternLength( int pattern, int len ) {
-  patternSettings[pattern].len = len - 1;
-}
-
-uint8_t PatternChannel( int pattern ) {
-  return patternSettings[pattern].channel + 1;
+// TODO: this should probably just be a constructor?
+SequencerState defaultSequencer() {
```

Comment:
Yeah I think so. A constructor and an init list, [I'm working on something similar here](https://github.com/winder/TSynth-Teensy4.1/blob/df4e86ec17db07a5ce09338eb3333bc8e48762e1/TSynth/VoiceGroup.cpp#L12).